src/rust/engine/ui/src/console_ui.rs

```diff
@@ -0,0 +1,222 @@
+use task_executor::Executor;
+use workunit_store::WorkUnitStore;
+
+pub struct ConsoleUI {
+  workunit_store: WorkUnitStore,
+  // While the UI is running, there will be an Instance present.
+  instance: Option<Instance>,
+}
+
+impl ConsoleUI {
+  pub fn new(workunit_store: WorkUnitStore) -> ConsoleUI {
+    ConsoleUI {
+      workunit_store,
+      instance: None,
+    }
+  }
+
+  fn default_draw_target() -> ProgressDrawTarget {
+    // NB: We render more frequently than we receive new data in order to minimize aliasing where a
+    // render might barely miss a data refresh.
+    ProgressDrawTarget::stderr_with_hz(Self::render_rate_hz() * 2)
+  }
+
+  ///
+  /// The number of times per-second that `Self::render` should be called.
+  ///
+  pub fn render_rate_hz() -> u64 {
+    10
+  }
+
+  pub fn write_stdout(&self, msg: &str) {
+    print!("{}", msg);
+  }
+
+  pub fn write_stderr(&self, msg: &str) {
+    if let Some(instance) = &self.instance {
+      instance.bars[0].println(msg);
+    } else {
+      eprintln!("{}", msg);
+    }
+  }
+
+  pub fn with_console_ui_disabled<F: FnOnce() -> T, T>(&self, f: F) -> T {
+    if let Some(instance) = &self.instance {
+      instance
+        .multi_progress
+        .set_draw_target(ProgressDrawTarget::hidden());
+      let res = f();
+      instance
+        .multi_progress
+        .set_draw_target(Self::default_draw_target());
+      res
+    } else {
+      f()
+    }
+  }
+
+  fn setup_bars(num_swimlanes: usize) -> (MultiProgress, Vec<ProgressBar>) {
+    let multi_progress_bars = MultiProgress::with_draw_target(Self::default_draw_target());
+
+    let bars = (0..num_swimlanes)
+      .map(|_n| {
+        let style = ProgressStyle::default_bar().template("{spinner} {wide_msg}");
+        multi_progress_bars.add(ProgressBar::new(50).with_style(style))
+      })
+      .collect();
+
+    (multi_progress_bars, bars)
+  }
+
+  fn get_label_from_heavy_hitters<'a>(
+    tasks_to_display: impl Iterator<Item = (&'a String, &'a Option<Duration>)>,
+  ) -> Vec<String> {
+    tasks_to_display
+      .map(|(label, maybe_duration)| {
+        let duration_label = match maybe_duration {
+          None => "(Waiting) ".to_string(),
+          Some(duration) => {
+            let duration_secs: f64 = (duration.as_millis() as f64) / 1000.0;
+            format!("{:.2}s ", duration_secs)
+          }
+        };
+        format!("{}{}", duration_label, label)
+      })
+      .collect()
+  }
+
+  ///
+  /// Updates all of the swimlane ProgressBars with new data from the WorkUnitStore. For this
+  /// method to have any effect, the `initialize` method must have been called first.
+  ///
+  /// *Technically this method does not do the "render"ing: rather, the `MultiProgress` instance
+  /// running on a background thread is drives rendering, while this method feeds it new data.
+  ///
+  pub fn render(&mut self) {
+    let instance = if let Some(i) = &mut self.instance {
+      i
+    } else {
+      return;
+    };
```

Comment:
I can't grok this. Is this unwrapping the `Option`?

Reply:
Yes. This code pattern-matches on `self.instance`, which is an `Option<T>`. If it is `Some`, it unwraps the inner value and sets the value of the local variable `instance` to it. If it is `None`, control flow exits the `render()` function entirely, so that local variable is moot.

Reply:
Oh, the `return` propagates to the entire function? I was thinking it would just "return" from the block. Interesting.